packages/flutter_tools/lib/src/extension/doctor.dart

```diff
@@ -0,0 +1,222 @@
+class ValidationResult implements Serializable {
+  /// Create a new [ValidationResponse].
+  ///
+  /// [messages], [type], and [name] must not be null.
+  /// If [messages] is not provided, it defaults to an empty list.
+  /// if [type] is not provided, it defaults to [ValidationType.installed].
+  const ValidationResult({
+    this.messages = const <ValidationMessage>[],
+    this.type = ValidationType.installed,
+    this.statusText,
+    @required this.name,
+  }) : assert(name != null),
+       assert(type != null),
+       assert(messages != null);
+
+  /// Create a new [ValidationResponse] from a json encoded object.
+  factory ValidationResult.fromJson(Map<String, Object> json) {
+    final String name = json['name'];
+    final List<ValidationMessage> messages = <ValidationMessage>[
+      for (Object value in json['messages'])
+        ValidationMessage.fromJson(value)
+    ];
+    final ValidationType type = ValidationType.fromJson(json['type']);
+    final String statusText = json['statusText'];
+    return ValidationResult(
+      name: name,
+      messages: messages,
+      type: type,
+      statusText: statusText,
+    );
+  }
+
+  /// The messages associated with this doctor diagnose.
+  ///
+  /// If not explicitly provided in the constructor this defaults to an empty
+  /// list. This is equivalent to claiming that there is nothing to diagnose in
+  /// the current environment.
+  final List<ValidationMessage> messages;
+
+  /// The overal validation summary.
+  ///
+  /// If not explicitly provided in the constructor this defaults tp
+  /// [ValidationType.installed]. This is eqivalent to claiming that there
+  /// is nothing wrong with the current environment.
+  ///
+  /// If there are no associated [messages] this instead means that there are
+  /// no applicable checks.
+  final ValidationType type;
+
+  /// The human readable name of the validator.
+  ///
+  /// Some interfaces, including the command line , will display the human
+  /// readable name next to the doctor results.
+  ///
+  /// Example: 'Android Studio', for validation of Android Studio installation.
+  final String name;
+
+  /// An optional high level summary.
+  ///
+  /// This should generally only be provided if a validation message came back
+  /// negative.
+  final String statusText;
+
+  @override
+  Map<String, Object> toJson() {
+    return <String, Object>{
+      'messages': <Object>[
+        for (ValidationMessage message in messages)
+          message.toJson()
+      ],
+      'type': type.toJson(),
+      'name': name,
+      'statusText': statusText,
+    };
+  }
+
+  @override
+  String toString() => json.encode(toJson());
+}
+
+/// Functionality related to diagnosing problems with a project or development
+/// environment.
+abstract class DoctorDomain extends Domain {
+
+  /// Diagnose problems with the user's environment.
+  Future<ValidationResult> diagnose(Map<String, Object> arguments);
```

Comment:
Not clear whether we should pass through arbitrary arguments or map it to something typed first. dynamic is better for forwards/backwards compatibility.